### PR TITLE
Improve docs for `struct_lint_level` function.

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @RalfJung 

Does this answer your questions?

### CODE REVIEW

File: compiler/rustc_middle/src/lint.rs [hunk 1]
```diff
@@ -274,6 +274,44 @@ pub fn explain_lint_level_source(
...
+///
+/// ## `decorate` signature
+///
+/// Signature of `decorate` may be confusing at first, for instance what's the
+/// point of returning `&'b mut DiagnosticBuilder<'a, ()>` if the original can
+/// be used instead?
+/// ```ignore pseudo-code
+/// _ = decorate(&mut diag);
+/// /* use `diag` here again */
+/// ```
```
Comment (RalfJung) [inline compiler/rustc_middle/src/lint.rs:295]: This is written from the perspective of whoever implements struct_lint_level, but the doc comment should be written for people that *call* `struct_lint_level`.
```suggestion
/// The return value of `decorate` is ignored by this function.So what is the
/// point of returning `&'b mut DiagnosticBuilder<'a, ()>`?
```

**Review by RalfJung — COMMENTED** [review]
The comment looks good, thanks! I left some suggestions.

There are a bunch of other functions also talking such a closure though. If I am calling `tcx.struct_span_lint_hir` I will still be confused. Could you add a comment to all of them just briefly stating that the return value is ignored and pointing to `struct_lint_level`?

**Comment by WaffleLapkin** [comment]
I've applied the suggestions from review, this can be reviewed again.

File: compiler/rustc_middle/src/lint.rs [hunk 2]
```diff
@@ -274,6 +274,39 @@ pub fn explain_lint_level_source(
...
+/// - `LintContext::lookup`
+///
+/// ## `decorate` signature
+///
+/// The return value of `decorate` is ignored by this function. So what is the
+/// point of returning `&'b mut DiagnosticBuilder<'a, ()>`?
+///
+/// There are 2 reasons for this signature.
+///
+/// First off all, it prevents accidental use of `.emit()` -- it's clear that the
```
Comment (RalfJung) [inline compiler/rustc_middle/src/lint.rs:294]:
```suggestion
/// First of all, it prevents accidental use of `.emit()` -- it's clear that the
```

**Comment by RalfJung** [comment]
Looks good, thanks a lot!
@bors r+ rollup

**Pull request merged**